File: data/Demon_matlab/NarrowBandAnalysis2.m
function[NBSNR, NBfreq] = NarrowBandAnalysis2(mixseg, noiseseg, Fs, pars);

% Performs a narrowband analysis of the signal

S = abs(fft(mixseg)).^2; % Spectral level
S = S(1:floor(length(S)/2));
Sn = abs(fft(noiseseg)).^2; % Spectral level
Sn = Sn(1:floor(length(Sn)/2));
f = Fs*[0:length(S)/2-1]/length(S);

% Whitening
% Snorm = S./Sn;
[Snorm] = normalise5(Snorm', pars.narrowband.N, pars.narrowband.G, 4);

%% Finding maximum peak above minimum frequency

% Keeping only frequency components within the set frequency window
els = find(f > pars.narrowband.minfreq & f < pars.narrowband.maxfreq);
Snorm = Snorm(els);
f = f(els);

% Finding the N strongest
[SnormSorted, I] = sort(Snorm);
NBfreq = sort(f(I(end:-1:end+1-pars.narrowband.NumLines)));
NBSNR = 10*log10(SnormSorted(end:-1:end+1-pars.narrowband.NumLines));

return

File: data/Demon_analysis/NarrowBandAnalysis2.py
import numpy as np
import scipy.signal as signal

def narrowband_analysis(mixseg, noiseseg, Fs, pars):
    """
    Performs a narrowband analysis of the signal.

    Parameters:
        mixseg (numpy array): Mixed signal segment.
        noiseseg (numpy array): Noise signal segment.
        Fs (int): Sampling frequency.
        pars (dict): Parameters dictionary.

    Returns:
        NBSNR (numpy array): Narrowband SNR values.
        NBfreq (numpy array): Frequencies of detected narrowband signals.
    """
    S = np.abs(np.fft.fft(mixseg)) ** 2  # Spectral level
    S = S[:len(S)//2]
    Sn = np.abs(np.fft.fft(noiseseg)) ** 2  # Noise spectral level
    Sn = Sn[:len(Sn)//2]

    f = Fs * np.arange(len(S)) / len(S)  # Frequency vector

    # Whitening
    Snorm = S / Sn
    Snorm = signal.medfilt(Snorm, kernel_size=pars["narrowband"]["N"])

    # Keeping only frequency components within the frequency window
    valid_indices = (f > pars["narrowband"]["minfreq"]) & (f < pars["narrowband"]["maxfreq"])
    Snorm = Snorm[valid_indices]
    f = f[valid_indices]

    # Finding the N strongest peaks
    sorted_indices = np.argsort(Snorm)
    NBfreq = np.sort(f[sorted_indices[-pars["narrowband"]["NumLines"]:]])
    NBSNR = 10 * np.log10(Snorm[sorted_indices[-pars["narrowband"]["NumLines"]:]])

    return NBSNR, NBfreq
